### PR TITLE
Grouping by frequency and resampling

### LINKED ISSUE #6255: [FEA] Support time-frequency grouping for Grouper
**Is your feature request related to a problem? Please describe.**

The `Grouper` in cuDF does not support most of the feature of `pd.Grouper`, including the `freq` parameter for grouping by a time frequency (e.g. daily, hourly).

ref: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Grouper.html

**Describe the solution you'd like**

Support `freq` and other features of `pd.Grouper` in cuDF

Also, cuDF `Grouper` is essentially undocumented and needs documentation support. 

**Describe alternatives you've considered**

Currently, grouping by a time period requires clunky code like
```
one_day = np.datetime64(1, 'D').astype('datetime64[ns]').astype('int64') 

df['from_day'] = df['starttime'].astype('int64') // one_day

df_out = df.groupby(by=["from_station_id", "from_day"])
```

### PR DESCRIPTION
**Pull request by shwina**
Closes #6255, #8416 

This PR implements two related features:

1. Grouping by a frequency via the `freq=` argument to `cudf.Grouper`
2. and time-series resampling via the `.resample()` API

Either operation results in ` _Resampler` object that represents the data resampled into "bins" of a particular frequency. The following operations are supported on resampled data:

1. Aggregations such as `min()` and `max()`, performed bin-wise
2. `ffill()` and `bfill()` methods: forward and backward filling in the case of upsampling data
3. `asfreq()`: returns the resampled data as a Series or DataFrame()

These are all best understood by example:

First, we create a time series with 1 minute intervals:

```python
>>> index = cudf.date_range(start="2001-01-01", periods=10, freq="1T")
>>> sr = cudf.Series(range(10), index=index)
>>> sr
2001-01-01 00:00:00    0
2001-01-01 00:01:00    1
2001-01-01 00:02:00    2
2001-01-01 00:03:00    3
2001-01-01 00:04:00    4
2001-01-01 00:05:00    5
2001-01-01 00:06:00    6
2001-01-01 00:07:00    7
2001-01-01 00:08:00    8
2001-01-01 00:09:00    9
dtype: int64
````
Downsampling to 3 minute intervals, followed by a "sum" aggregation:

```python
>>> sr.resample("3T").sum()  # equivalently, sr.groupby(cudf.Grouper(freq="3T")).sum()
2001-01-01 00:00:00     3
2001-01-01 00:03:00    12
2001-01-01 00:06:00    21
2001-01-01 00:09:00     9
dtype: int64
````

Upsampling to 30 second intervals:

```python
>>> sr.resample("30s").asfreq()
2001-01-01 00:00:00    0.0
2001-01-01 00:00:30    NaN
2001-01-01 00:01:00    1.0
2001-01-01 00:01:30    NaN
2001-01-01 00:02:00    2.0
2001-01-01 00:02:30    NaN
2001-01-01 00:03:00    3.0
2001-01-01 00:03:30    NaN
2001-01-01 00:04:00    4.0
2001-01-01 00:04:30    NaN
2001-01-01 00:05:00    5.0
2001-01-01 00:05:30    NaN
2001-01-01 00:06:00    6.0
2001-01-01 00:06:30    NaN
2001-01-01 00:07:00    7.0
2001-01-01 00:07:30    NaN
2001-01-01 00:08:00    8.0
2001-01-01 00:08:30    NaN
2001-01-01 00:09:00    9.0
Freq: 30S, dtype: float64
```

Upsampling to 30 second intervals, followed by a forward fill:

```python
>>> sr.resample("30s").ffill()
2001-01-01 00:00:00    0
2001-01-01 00:00:30    0
2001-01-01 00:01:00    1
2001-01-01 00:01:30    1
2001-01-01 00:02:00    2
2001-01-01 00:02:30    2
2001-01-01 00:03:00    3
2001-01-01 00:03:30    3
2001-01-01 00:04:00    4
2001-01-01 00:04:30    4
2001-01-01 00:05:00    5
2001-01-01 00:05:30    5
2001-01-01 00:06:00    6
2001-01-01 00:06:30    6
2001-01-01 00:07:00    7
2001-01-01 00:07:30    7
2001-01-01 00:08:00    8
2001-01-01 00:08:30    8
2001-01-01 00:09:00    9
Freq: 30S, dtype: int64
```

### CODE REVIEW

**Comment by shwina** [comment]
Some quick and dirty benchmarking:

Speedups for downsampling:

downsampling

Speedups for upsampling:

upsampling

Benchmarking  script:

<details>

```python
import timeit

import pandas as pd
import cudf
import numpy as np
import cupy as cp

def bench_pandas_resample(n, kind="downsample"):
    if kind == "downsample":
        init_freq = "1T"
        resampled_freq = "3T"
    else:
        init_freq = "3T"
        resampled_freq = "1T"
    
    rng = pd.date_range("2001-01-01", periods=n, freq=init_freq)
    sr = pd.Series(np.random.randint(0, 100, len(rng)), index=rng)
    
    nrep = 10
    sr.resample(resampled_freq).mean()
    t1 = timeit.default_timer()
    for i in range(nrep):
        sr.resample(resampled_freq).mean()
    t2 = timeit.default_timer()
    return (t2 - t1) / nrep

def bench_cudf_resample(n, kind="downsample"):
    if kind == "downsample":
        init_freq = "1T"
        resampled_freq = "3T"
    else:
        init_freq = "3T"
        resampled_freq = "1T"
        
    rng = cudf.date_range("2001-01-01", periods=n, freq=init_freq)
    sr = cudf.Series(cp.random.randint(0, 100, len(rng)), index=rng)
    
    nrep = 10
    sr.resample(resampled_freq).mean()
    t1 = timeit.default_timer()
    for i in range(nrep):
        sr.resample(resampled_freq).mean()
    t2 = timeit.default_timer()
    return (t2 - t1) / nrep

import matplotlib.pyplot as plt

speedups = []
sizes = (10_000, 100_000, 1_000_000, 10_000_000)

for size  in  sizes:
    speedups.append(bench_pandas_resample(size, kind="downsample") / bench_cudf_resample(size, kind="downsample"))

plt.bar(range(len(sizes)), speedups)


plt.bar(range(len(sizes)), speedups)
plt.xticks(range(len(sizes)), [str(x) for x in  sizes])
plt.xlabel("DataFrame size")
plt.ylabel("Speedup")

```

</details>

File: python/cudf/cudf/core/column_accessor.py [hunk 1]
```diff
@@ -537,6 +537,27 @@ def rename_column(x):
 
         return self.__class__(ca)
 
+    def droplevel(self, level):
+        # drop the nth level
+        if self.nlevels == 1:
+            raise TypeError("Cannot drop the only remaining level")
```
Comment (shwina) [inline python/cudf/cudf/core/column_accessor.py:543]: Needs to be a `ValueError` instead.

File: python/cudf/cudf/core/dataframe.py [hunk 2]
```diff
@@ -821,6 +822,34 @@ def _from_data(
             out.columns = columns
         return out
 
+    def _align_to_index(
```
Comment (shwina) [inline python/cudf/cudf/core/dataframe.py:825]: Move this to `IndexedFrame` perhaps?

**Comment by vuule** [comment]
What is the unit on the `speedup` axis?

**Comment by shwina** [comment]
`cudf_runtime / pandas_runtime`

**Comment by vuule** [comment]
> `pandas_runtime / cudf_runtime`

Does that mean that cudf is a lot slower than pandas for small dataframes (up to 5x? it's hard to say)?

**Comment by shwina** [comment]
Yes -- unfortunately, that's the case :-( Ideally, I'd like to get this initial implementation in by 21.12; we're still thinking about opportunities for improving performance.

**Review by isVoid — COMMENTED** [review]
I'm just half way done. I'm already learning a lot in the process reading it. It's a bit hard to follow in the middle of `_ResampleGrouping` and I think it's best to step through it with an example. I'll continue tm.

**Comment by harrism** [comment]
@shwina can you write a description for this PR? The description ends up in the changelog, so "fixes #" is not descriptive enough IMO. I saw the speedup graphs, but I don't really know what this PR is about other than the title, so it's hard to appreciate them.  (BTW if you turn off dark mode you can see the axes.)

**Comment by harrism** [comment]
Removed cpp-codeowners review requirement since there seem to be no C++ changes.

**Comment by ajschmidt8** [comment]
Removing `ops-codeowners` from the required reviews since it doesn't seem there are any file changes that we're responsible for. Feel free to add us back if necessary.

File: python/cudf/cudf/core/join/join.py [hunk 3]
```diff
@@ -339,6 +347,21 @@ def _merge_results(self, left_result: Frame, right_result: Frame) -> Frame:
 
         return result
 
+    @property
+    def _result_has_multiindex_columns(self):
```
[the remark this answers is not in the document]
Comment (brandon-b-miller) [reply python/cudf/cudf/core/join/join.py:351]: Since this can be determined from the merge configuration I might suggest just computing it on init and storing it as an attribute, or as part of the merge parameter preprocessing step

**Comment by brandon-b-miller** [comment]
Still reviewing this - I am building this branch to play around live a bit.

File: python/cudf/cudf/core/dataframe.py [hunk 4]
```diff
@@ -821,6 +822,34 @@ def _from_data(
...
+            return self
+
+        lhs = self.copy(deep=False)
+        rhs = cudf.DataFrame(index=as_index(index))
+
+        rng_col_id = str(uuid4())
+        if how == "left":
+            lhs[rng_col_id] = column.arange(len(lhs))
+        else:
+            rhs[rng_col_id] = column.arange(len(rhs))
```
Comment (brandon-b-miller) [inline python/cudf/cudf/core/dataframe.py:840]: It seems like this gets allocated and then unused in the `how=outer` case. If that's the case this might be worth addressing since it is the default case.

File: python/cudf/cudf/core/resample.py [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+        # bin all the values. It's OK if we generate more labels than
+        # we need, as we remove any unused labels below
+        end += offset
+
+        # generate the labels for binning the key column:
+        bin_labels = cudf.date_range(start=start, end=end, freq=freq,)
+
+        # We want the (resampled) column of timestamps in the result
+        # to have a resolution closest to the resampling
+        # frequency. For example, if resampling from '1T' to '1s', we
```
Comment (brandon-b-miller) [inline python/cudf/cudf/core/resample.py:160]: since this is a place we diverge from pandas we should probably be loud about this in the docs, and note the limitations/future direction we settled on here

Comment (brandon-b-miller) [inline python/cudf/cudf/core/resample.py:160]: since this is a place we diverge from pandas we should probably be loud about this in the docs, and note the limitations/future direction we settled on here

File: python/cudf/cudf/core/resample.py [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+    def asfreq(self):
+        return self.obj._align_to_index(
+            self.grouping.bin_labels,
+            how="right",
+            sort=False,
+            allow_non_unique=True,
+        )
+
+    def _scan_fill(self, method: str, limit: int) -> DataFrameOrSeries:
+        # TODO: can this be more efficient?
```
[the remark this answers is not in the document]
Comment (shwina) [reply python/cudf/cudf/core/resample.py:51]: Any luck with this? (I note that this is an example of a situation where we can benefit from caching the join hash table, where maybe the second join could be really cheap).

File: python/cudf/cudf/core/resample.py (hunk 5 above)
Comment (shwina) [reply python/cudf/cudf/core/resample.py:160]: I'm just afraid that a warning here might look a bit too scary. We support multiple frequencies and are taking advantage of that, where Pandas cannot. Maybe calling out this behaviour in the docs is better?

File: python/cudf/cudf/core/resample.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (brandon-b-miller) [reply python/cudf/cudf/core/resample.py:51]: jumped to something else temporarily but will rotate back to this this afternoon

File: python/cudf/cudf/tests/test_datetime.py [hunk 7]
```diff
@@ -1452,7 +1452,14 @@ def test_is_month_start(data, dtype):
...
     {"months": 3, "years": 1},
-    {"hours": 10, "days": 57, "nanoseconds": 3},
+    pytest.param(
+        {"hours": 10, "days": 57, "nanoseconds": 3},
+        marks=pytest.mark.xfail(
+            True,
+            reason="Pandas ignoring nanoseconds component. "
+            "https://github.com/pandas-dev/pandas/issues/44393",
+        ),
+    ),
```
Comment (shwina) [inline python/cudf/cudf/tests/test_datetime.py:1462]: @isVoid I'm not sure how/why we were passing these tests before, but it looked to me like we were broken in a similar way to Pandas.

File: python/cudf/cudf/core/resample.py (hunk 5 above)
Comment (brandon-b-miller) [reply python/cudf/cudf/core/resample.py:160]: oh sorry, I don't mean a warning- I just mean moving some info about this to the docstring.

File: python/cudf/cudf/core/tools/datetimes.py [hunk 8]
```diff
@@ -881,13 +900,12 @@ def date_range(
...
-            start=start.value.astype("int64"),
-            stop=end_estim.astype("int64"),
-            num=periods,
-        )
+        # If `offset` is fixed frequency, we generate a range of
+        # treating `start`, `stop` and `step` as ints:
+        stop = end_estim.astype("int64")
+        start = start.value.astype("int64")
+        step = int(_offset_to_nanoseconds_lower_bound(offset))
+        arr = cp.arange(start=start, stop=stop, step=step)
```
Comment (shwina) [inline python/cudf/cudf/core/tools/datetimes.py:908]: FYI @isVoid - let's sync up offline about these changes.

File: python/cudf/cudf/core/resample.py (hunk 5 above)
Comment (shwina) [reply python/cudf/cudf/core/resample.py:160]: Gotcha! Updated  the docs to have a `Notes` section with this info

File: python/cudf/cudf/core/resample.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (shwina) [reply python/cudf/cudf/core/resample.py:51]: No worries -- would you be OK with possibly adding it on in a follow-up?

[the remark this answers is not in the document]
Comment (brandon-b-miller) [reply python/cudf/cudf/core/resample.py:51]: 👍

**Comment by shwina** [comment]
rerun tests

File: python/cudf/cudf/core/tools/datetimes.py (hunk 8 above)
Comment (isVoid) [reply python/cudf/cudf/core/tools/datetimes.py:908]: I don't think this regression is tested in the test cases, should we add it? This snippet you sent me earlier:
```python
cudf.date_range('2001-01-01 00:00:00.000000', '2001-01-01 00:00:00.000010', freq='3us')
```
can be tested against pandas.

**Comment by isVoid** [comment]
Currently when I run it with any operation, because `IndexedFrame.sort_values` calls `take`, I get a `FutureWarning` which is quite confusing to user:
```python
In [10]: gdf.groupby(cudf.Grouper(key="Publish date", freq="1h", label="left", closed="left")).sum()
/raid/wangm/dev/rapids/cudf/python/cudf/cudf/core/frame.py:3076: FutureWarning: keep_index is deprecated and will be removed in the future.
  FutureWarning,
Out[10]: 
                       ID Price
Publish date                   
2000-01-01 12:00:00     1    30
2000-01-01 13:00:00  <NA>  <NA>
2000-01-01 14:00:00  <NA>  <NA>
...
```

File: python/cudf/cudf/core/join/join.py [hunk 9]
```diff
@@ -310,8 +316,25 @@ def _merge_results(self, left_result: Frame, right_result: Frame) -> Frame:
...
                 del right_names[name]
 
-        # Assemble the data columns of the result:
-        data = left_result._data.__class__()
+        # determine if the result has multiindex columns.  The result
+        # of a join has a MultiIndex as its columns if:
+        # - both the `lhs` and `rhs` have a MultiIndex columns
+        # OR
+        # - either one of `lhs` or `rhs` have a MultiIndex columns,
+        #   and the other is empty (i.e., no columns)
```
Comment (brandon-b-miller) [inline python/cudf/cudf/core/join/join.py:324]: Dont we also get a multiindex if `left_index=True` and `right_on` is a list of the same number of columns, or the other way around?

Comment (shwina) [reply python/cudf/cudf/core/join/join.py:324]: This comment is not about MultiIndex `index`,  but rather MultiIndex `columns`. I wonder if there's other language we can use to make the distinction more clear? We use `MultiColumn` informally but that term would be unfamiliar to newcomers.

Comment (brandon-b-miller) [reply python/cudf/cudf/core/join/join.py:324]: Oh I see, thanks for clarifying 👍

File: python/cudf/cudf/core/resample.py [hunk 10]
```diff
@@ -0,0 +1,360 @@
...
+
+class _ResampleGrouping(_Grouping):
+
+    bin_labels: cudf.Index
+
+    def _handle_frequency_grouper(self, by):
+        # if `by` is a time frequency grouper, we bin the key column
+        # using bin intervals specified by `by.freq`, then use *that*
+        # as the groupby key
+        import cudf._lib.labeling
```
Comment (brandon-b-miller) [inline python/cudf/cudf/core/resample.py:90]: can this be removed or moved to module level?

File: python/cudf/cudf/core/resample.py [hunk 11]
```diff
@@ -0,0 +1,360 @@
...
+            closed = "right" if closed is None else closed
+        else:
+            label = "left" if label is None else label
+            closed = "left" if closed is None else closed
+
+        # determine the key column
+        if by.key is None and by.level is None:
+            # then assume that the key is the index of `self._obj`:
+            if not isinstance(self._obj.index, cudf.DatetimeIndex):
+                raise TypeError(
```
Comment (brandon-b-miller) [inline python/cudf/cudf/core/resample.py:118]: Couldn't we remove this instance check/error case and it would still throw almost the same error once it made it down to the check on line 133?

File: python/cudf/cudf/core/index.py [hunk 12]
```diff
@@ -2329,6 +2329,7 @@ def __init__(
             data = column.as_column(data)
             data.dtype.closed = closed
 
+        self.closed = closed
```
Comment (vyasr) [inline python/cudf/cudf/core/index.py:2332]: Was this just a bug before? I see that `cut` at least always expected this attribute to exist.

File: python/cudf/cudf/core/groupby/groupby.py [hunk 13]
```diff
@@ -1303,29 +1300,42 @@ def apply(self, func):
...
+        else:
+            return cudf.core.resample.DataFrameResampler(obj, by=by)
+
+    if isinstance(obj, cudf.Series):
+        return SeriesGroupBy(obj, by=by, **kwargs)
+    else:
+        return DataFrameGroupBy(obj, by=by, **kwargs)
+
+
+# TODO: should we define this as a dataclass instead?
```
[the remark this answers is not in the document]
Comment (vyasr) [reply python/cudf/cudf/core/groupby/groupby.py:1318]: Ah I see that this is actually part of the pandas API. Never mind then.

File: python/cudf/cudf/core/indexed_frame.py [hunk 14]
```diff
@@ -526,3 +532,208 @@ def _n_largest_or_smallest(self, largest, n, columns, keep):
...
+        how: str = "outer",
+        sort: bool = True,
+        allow_non_unique: bool = False,
+    ) -> T:
+        index = cudf.core.index.as_index(index)
+
+        if self.index.equals(index):
+            return self
+        if not allow_non_unique:
+            if len(self) != len(self.index.unique()) or len(index) != len(
```
Comment (vyasr) [inline python/cudf/cudf/core/indexed_frame.py:548]: I think you can use `index.is_unique`.

File: python/cudf/cudf/core/indexed_frame.py [hunk 15]
```diff
@@ -526,3 +532,208 @@ def _n_largest_or_smallest(self, largest, n, columns, keep):
...
+                index.unique()
+            ):
+                raise ValueError("Cannot align indices with non-unique values")
+
+        lhs = cudf.DataFrame._from_data(self._data, index=self.index)
+        rhs = cudf.DataFrame(index=index)
```
Comment (vyasr) [inline python/cudf/cudf/core/indexed_frame.py:554]: Any reason this can't also use `_from_data`?

File: python/cudf/cudf/core/indexed_frame.py [hunk 16]
```diff
@@ -526,3 +532,208 @@ def _n_largest_or_smallest(self, largest, n, columns, keep):
...
+
+        tmp_col_id = str(uuid4())
+        if how == "left":
+            lhs[tmp_col_id] = arange(len(lhs))
+        elif how == "right":
+            rhs[tmp_col_id] = arange(len(rhs))
+
+        result = lhs.join(rhs, how=how, sort=sort)
+        if how == "left" or how == "right":
```
Comment (vyasr) [inline python/cudf/cudf/core/indexed_frame.py:563]:
```suggestion
        if how in ("left", "right"):
```

File: python/cudf/cudf/core/indexed_frame.py [hunk 17]
```diff
@@ -526,3 +532,208 @@ def _n_largest_or_smallest(self, largest, n, columns, keep):
...
+        if not allow_non_unique:
+            if len(self) != len(self.index.unique()) or len(index) != len(
+                index.unique()
+            ):
+                raise ValueError("Cannot align indices with non-unique values")
+
+        lhs = cudf.DataFrame._from_data(self._data, index=self.index)
+        rhs = cudf.DataFrame(index=index)
+
+        tmp_col_id = str(uuid4())
```
Comment (vyasr) [inline python/cudf/cudf/core/indexed_frame.py:556]: A comment on why this temp col is necessary (for sorting) would be helpful. A more meaningful name e.g. sorted_index_col might also be helpful.

File: python/cudf/cudf/core/indexed_frame.py [hunk 18]
```diff
@@ -526,3 +532,208 @@ def _n_largest_or_smallest(self, largest, n, columns, keep):
...
+            lhs[tmp_col_id] = arange(len(lhs))
+        elif how == "right":
+            rhs[tmp_col_id] = arange(len(rhs))
+
+        result = lhs.join(rhs, how=how, sort=sort)
+        if how == "left" or how == "right":
+            result = result.sort_values(tmp_col_id)
+            del result[tmp_col_id]
+
+        result = self.__class__._from_data(result._data, index=result.index)
```
Comment (vyasr) [inline python/cudf/cudf/core/indexed_frame.py:567]: Just to make sure I understand, for a 1D frame (Series) this function is going to make a 2-col DataFrame for the purposes of sort order but then just extract the index-sorted original column back out to reconstruct, right? I don't really like that we have to refer to DataFrame in this class, but I don't know if there's a great alternative since moving the impl of some of these methods to Frame would expose Series methods that shouldn't exist.

This is probably out of scope for this PR, these types of logical circular references exist in multiple places in cudf, but it's something I'd like to address eventually. We should think a bit about whether we can/should move all Series methods that require conversion to DataFrame to the Series level rather than in IndexedFrame. Otherwise it's bound to lead to inconsistent implementations at worst, and less efficient implementations at best.

File: python/cudf/cudf/core/join/join.py [hunk 19]
```diff
@@ -354,7 +377,9 @@ def _sort_result(self, result: Frame) -> Frame:
                 sort_order = result._get_sorted_inds(
                     list(_coerce_to_tuple(self.on))
                 )
-            return result._gather(sort_order, keep_index=False)
+            return result._gather(
+                sort_order, keep_index=False, check_bounds=False
```
Comment (vyasr) [inline python/cudf/cudf/core/join/join.py:381]: I wonder if there are other places that we might leverage this optimization (probably out of scope for this PR, just want to plant the idea in your head).

File: python/cudf/cudf/core/tools/datetimes.py [hunk 20]
```diff
@@ -859,15 +877,16 @@ def date_range(
...
+    # range) below, but don't always use it.  We do this to ensure
+    # that the appropriate OverflowError is raised by Pandas in case
+    # of overflow.
     # FIXME: when `end_estim` is out of bound, but the actual `end` is not,
     # we shouldn't raise but compute the sequence as is. The trailing overflow
     # part should get trimmed at the end.
     end_estim = (
         pd.Timestamp(start.value)
-        + (periods - 1) * offset._maybe_as_fast_pandas_offset()
+        + periods * offset._maybe_as_fast_pandas_offset()
```
Comment (vyasr) [inline python/cudf/cudf/core/tools/datetimes.py:889]: Was this a bug?

File: python/cudf/cudf/core/resample.py [hunk 21]
```diff
@@ -0,0 +1,360 @@
...
+)
+
+
+class _Resampler(GroupBy):
+
+    grouping: "_ResampleGrouping"
+
+    def __init__(self, obj, by, axis=None, kind=None):
+        if not isinstance(by, Grouper) and by.freq:
+            # TODO
```
Comment (vyasr) [inline python/cudf/cudf/core/resample.py:23]: ?

File: python/cudf/cudf/core/resample.py [hunk 22]
```diff
@@ -0,0 +1,360 @@
...
+        # frequency. For example, if resampling from '1T' to '1s', we
+        # want the resulting timestamp column to by of dtype
+        # 'datetime64[s]'.  libcudf requires the bin labels and key
+        # column to have the same dtype, so we compute a `result_type`
+        # and cast them both to that type.
+        try:
+            from cudf.core.tools.datetimes import (
+                _offset_alias_to_code,
+                _unit_dtype_map,
+            )
```
Comment (vyasr) [inline python/cudf/cudf/core/resample.py:168]: Why the local import? These seem fine to import at module level.

File: python/cudf/cudf/core/resample.py [hunk 23]
```diff
@@ -0,0 +1,360 @@
...
+        # replace self._key_columns with the binned key column:
+        self._key_columns = [
+            bin_labels._gather(bin_numbers, check_bounds=False)._column.astype(
+                result_type
+            )
+        ]
+
+
+# NOTE: this function is vendored from Pandas
+def _get_timestamp_range_edges(
```
Comment (vyasr) [inline python/cudf/cudf/core/resample.py:215]: @brandon-b-miller @isVoid the last time I spoke with @shwina I mentioned that I might actually prefer to use pandas internal APIs rather than vendor them. Here's how I see the two options:
1. We vendor the functions. The code will always work, even if pandas changes the signature of these functions. However, if pandas changes the behavior of these functions without changing the signature, our behavior could silently diverge from pandas unless our test coverage is good enough to catch all edge cases (which is basically impossible given how many edge cases there are in pandas APIs).
2. We use pandas internal APIs. If pandas changes the signature of these functions, cudf immediately fails, but it fails loudly. Meanwhile, we are automatically kept in sync with pandas behavior changes.

It seems to me that if our goal is to match all behavior of pandas, we should prefer option (2) even if it means using pandas internals since that's effectively what we're doing anyway. Thoughts?

**Review by vyasr — CHANGES_REQUESTED** [review]

File: python/cudf/cudf/core/resample.py (hunk 23 above)
Comment (brandon-b-miller) [reply python/cudf/cudf/core/resample.py:215]: I think it ends up being conservation of misery at some point because even if we import the functions so that we can keep up to date with any changes, they could be fully moved somewhere else at some point or even removed completely. So we trade the maintenance burden of keeping them up to date ourselves with the risk of having to go find them or even vendor them later anyways if they disappear and aren't replaced with something suitable for our uses. 

On the other hand vendoring them creates a "tribal knowledge" situation that makes it harder for contributors to know how to maintain the codebase along with the divergence issue you mentioned. Maybe we can judge it on a case by case basis? In the case of `_get_timestamp_range_edges` it looks like the code has changed as little as 6 months ago but the function has lived in the file for almost a decade, based on the blame. I suppose given that I might lean towards importing it...

File: python/cudf/cudf/core/column_accessor.py [hunk 24]
```diff
@@ -537,6 +537,27 @@ def rename_column(x):
 
         return self.__class__(ca)
 
+    def droplevel(self, level):
```
[the remark this answers is not in the document]
Comment (shwina) [reply python/cudf/cudf/core/column_accessor.py:540]: Note that `ColumnAccessor` _is_ our implementation of "multi-level columns". The `.columns` attribute of a cudf.DataFrame doesn't return a `cudf.MultiIndex`, but rather a `pd.MultiIndex`. So operations like dropping a level of a multi-level column need to be either methods of `ColumnAccessor` or free functions that act on `ColumnAccessor`.

File: python/cudf/cudf/core/tools/datetimes.py (hunk 20 above)
Comment (isVoid) [reply python/cudf/cudf/core/tools/datetimes.py:889]: Yes, the changes below that replaces `linspace` with `arrange` also have to do with this.
See: https://github.com/rapidsai/cudf/pull/9178#discussion_r747815516

File: python/cudf/cudf/core/column_accessor.py [hunk 25]
```diff
@@ -537,6 +537,27 @@ def rename_column(x):
...
+        # drop the nth level
+        if self.nlevels == 1:
```
[the remark this answers is not in the document]
Comment (shwina) [reply python/cudf/cudf/core/column_accessor.py:542]: Removed!

File: python/cudf/cudf/core/groupby/groupby.py [hunk 26]
```diff
@@ -240,22 +239,20 @@ def agg(self, func):
             result = result.sort_index()
 
         if not _is_multi_agg(func):
-            if result.columns.nlevels == 1:
+            if result._data.nlevels <= 1:  # 0 or 1 levels
```
[the remark this answers is not in the document]
Comment (shwina) [reply python/cudf/cudf/core/groupby/groupby.py:242]: In either case, I believe we want to demote to a flat index here.

File: python/cudf/cudf/core/groupby/groupby.py [hunk 27]
```diff
@@ -1303,29 +1300,42 @@ def apply(self, func):
...
         self._key_columns = []
         self.names = []
-        # For transform operations, we want to filter out only the value
-        # columns that will be used. When part of the key columns are composed
-        # from columns in `obj`, these columns are not included in the value
-        # columns. This only happens when `by` is specified with
-        # column labels or `Grouper` object with `key` param. To avoid that
-        # external objects overlaps in names to `obj`, these column
-        # names are recorded separately in this list.
-        self._key_column_names_from_obj = []
```
[the remark this answers is not in the document]
Comment (shwina) [reply python/cudf/cudf/core/groupby/groupby.py:1328]: Yup.

File: python/cudf/cudf/core/column_accessor.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (vyasr) [reply python/cudf/cudf/core/column_accessor.py:540]: Ah of course. Yeah I guess we're stuck with this. I dislike the methods in ColumnAccessor that are used very sparingly since I'd like to thin out this class rather than augment it, but for now I think we live with the result.

File: python/cudf/cudf/core/tools/datetimes.py (hunk 20 above)
Comment (vyasr) [reply python/cudf/cudf/core/tools/datetimes.py:889]: Thanks for the context @isVoid!

File: python/cudf/cudf/core/index.py (hunk 12 above)
Comment (shwina) [reply python/cudf/cudf/core/index.py:2332]: I think so..

File: python/cudf/cudf/core/indexed_frame.py (hunk 14 above)
Comment (shwina) [reply python/cudf/cudf/core/indexed_frame.py:548]: 👍

File: python/cudf/cudf/core/indexed_frame.py (hunk 18 above)
Comment (shwina) [reply python/cudf/cudf/core/indexed_frame.py:567]: Correct -- I don't mind so much referring to subclasses in a base class. It's the reason we like dynamic languages like Python. I _don't_ like that we have to construct a DataFrame for this operation, where ideally some simpler, internal frame-like data structure would suffice.

File: python/cudf/cudf/core/join/join.py (hunk 19 above)
Comment (shwina) [reply python/cudf/cudf/core/join/join.py:381]: There's definitely a bunch.

File: python/cudf/cudf/core/resample.py (hunk 23 above)
Comment (shwina) [reply python/cudf/cudf/core/resample.py:215]: Vendoring also lets us tweak/optimize things for cuDF. For example, in these vendored functions, we can safely remove much of the timezone-related logic and not lose any cuDF functionality.

File: python/cudf/cudf/core/resample.py (hunk 10 above)
Comment (shwina) [reply python/cudf/cudf/core/resample.py]: Moved.

File: python/cudf/cudf/core/resample.py (hunk 11 above)
Comment (shwina) [reply python/cudf/cudf/core/resample.py:114]: Yup - you're right. Fixed!

File: python/cudf/cudf/core/column_accessor.py [hunk 28]
```diff
@@ -554,3 +575,14 @@ def _compare_keys(target: Any, key: Any) -> bool:
         if k1 != k2:
             return False
     return True
+
+
+def _remove_key_level(key: Any, level: int, detupleize=True) -> Any:
```
[the remark this answers is not in the document]
Comment (shwina) [reply python/cudf/cudf/core/column_accessor.py]: Removed.

File: python/cudf/cudf/core/resample.py (hunk 23 above)
Comment (shwina) [reply python/cudf/cudf/core/resample.py:201]: sync'd offline and we decided to leave these vendored for the moment, recognizing that there's no good solution.

**Review by brandon-b-miller — APPROVED** [review]

File: python/cudf/cudf/core/groupby/groupby.py [hunk 29]
```diff
@@ -1303,29 +1300,42 @@ def apply(self, func):
         return result
 
 
+def _get_groupby(obj, by, **kwargs):
```
[the remark this answers is not in the document]
Comment (shwina) [reply python/cudf/cudf/core/groupby/groupby.py]: Done

File: python/cudf/cudf/core/indexed_frame.py (hunk 18 above)
Comment (vyasr) [reply python/cudf/cudf/core/indexed_frame.py:566]: IMO while it's the kind of thing that dynamic languages _allow_, it is usually (but of course not always) a code smell. It's not that different from forward declarations or friends in C++; there's a time and place, but often it often indicates that there's a better design to be had. But that's something we can debate at a later point when we're not coming up on code freeze :)

**Review by vyasr — APPROVED** [review]

**Comment by quasiben** [comment]
@gpucibot  merge

**Review by quasiben — APPROVED** [review]

**Comment by shwina** [comment]
@gpucibot merge
